### PR TITLE
[hailtop]grouped type annotation

### PR DESCRIPTION
**Pull request by CDiaz96**

### CODE REVIEW

File: hail/python/hailtop/utils/utils.py [hunk 1]
```diff
@@ -107,7 +107,7 @@ def digits_needed(i: int) -> int:
     return 1 + digits_needed(i // 10)
 
 
-def grouped(n, ls):
+def grouped(n:int, ls:list[Any]) -> list:
```
Comment (jigold) [inline hail/python/hailtop/utils/utils.py:110]: A couple of things

1. Type annotations have an int after the colon
2. Types are all lowercase for things like int, str, float, bytes, but are in upper case (imported from Typing above) for things like Any, List, Dict, Tuple, Iterable, etc.
3. Think about what `grouped` is returning. Take a look at your test cases you wrote a week or two ago.

**Review by jigold — CHANGES_REQUESTED** [review]

**Comment by danking** [comment]
Also for names fo PR's we like to use `[component] blah blah`. So for this it would be `[hailtop] added type annotation`

File: hail/python/hailtop/utils/utils.py [hunk 2]
```diff
@@ -107,7 +108,7 @@ def digits_needed(i: int) -> int:
     return 1 + digits_needed(i // 10)
 
 
-def grouped(n: int, ls: List[Any]) -> List:
+def grouped(n: int, ls: List[Any]) -> Generator[Any]:
```
Comment (jigold) [inline hail/python/hailtop/utils/utils.py:111]: We want the type to be as specific as we know. What are we generating? i.e. what is the type of `group` in the code below?

**Review by jigold — CHANGES_REQUESTED** [review]

File: hail/python/hailtop/utils/utils.py (hunk 2 above)
Comment (CDiaz96) [reply hail/python/hailtop/utils/utils.py:111]: We are generating a list that could have either integers or strings.

File: hail/python/hailtop/utils/utils.py [hunk 3]
```diff
@@ -1,4 +1,5 @@
-from typing import Callable, TypeVar, Awaitable, Optional, Type, List, Dict, Iterable, Tuple
+from email.generator import Generator
+from typing import Any, Callable, TypeVar, Awaitable, Optional, Type, List, Dict, Iterable, Tuple
```
Comment (jigold) [inline hail/python/hailtop/utils/utils.py:2]: The first import is not needed. We want the Generator type from `typing`. You can just add it in Alphabetical order to the line below.

File: hail/python/hailtop/utils/utils.py (hunk 2 above)
Comment (jigold) [reply hail/python/hailtop/utils/utils.py:111]: Correct, so what should `Any` be in this case then?

**Review by jigold — CHANGES_REQUESTED** [review]

**Comment by CDiaz96** [comment]
Looking through the google doc I read Dan mentioning Iterators as returns types and since this is a list we are iterating through, I thought I should test it to see if the return type for Iterable would work and it did pass.

File: hail/python/hailtop/utils/utils.py [hunk 4]
```diff
@@ -107,7 +107,7 @@ def digits_needed(i: int) -> int:
     return 1 + digits_needed(i // 10)
 
 
-def grouped(n, ls):
+def grouped(n: int, ls: List[Any]) -> Iterable:
```
Comment (danking) [inline hail/python/hailtop/utils/utils.py:110]: Let's go with `Iterable[Any]`. `Iterable` is just like `List`: it's an iterable *of something*.

**Review by danking — CHANGES_REQUESTED** [review]

File: hail/python/hailtop/utils/utils.py (hunk 4 above)
Comment (CDiaz96) [reply hail/python/hailtop/utils/utils.py]: Okay. Yes, I went ahead and added it.

**Review by danking — APPROVED** [review]
nice

**Pull request merged**